portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -1,143 +1,159 @@
-    */
+  // Constants for information to be put into response of a GET request.
+  private static final String IS_USER_LOGGED_IN = "loggedIn"; 
+  private static final String REDIRECT_URL = "redirectUrl";
+  private static final String HOME_URL = "/"; 
+
+  // Constant for fetching quotes from Datastore
+  private static final int DEFAULT_NUM_QUOTES = 5;
+  
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    Query query = new Query(QUOTE).addSort(TIMESTAMP, SortDirection.DESCENDING);
-    PreparedQuery quotes = datastore.prepare(query);
+    response.setContentType("application/json");
 
-    // Get the parameter for number to display, with 5 as the default value.
-    int numQuoteToDisplay = getNumQuoteDisplayed(request, DEFAULT_NUM_QUOTES);
-   
-    // Add each quote to the quote list whose content will be written as 
-    // the response from the servlet. 
-    List<Quote> quoteList = new ArrayList<>();
-    quotes.asList(FetchOptions.Builder.withLimit(numQuoteToDisplay))
-        .forEach((quoteEntity) -> {
-          quoteList.add(extractQuoteFromEntity(quoteEntity));
-        });
-    
-    // Put the thread to sleep for a short period of time.
-    // This adds buffer time for querying Datastore. 
-    try {
-      Thread.sleep(200);
-    } catch (Exception e) {
-      System.out.println(e);
+    UserService userService = UserServiceFactory.getUserService();
+
+    Map<String, String> responseMap = new HashMap<String, String>();
+    User currentUser = userService.getCurrentUser(); 
+
+    if (userService.isUserLoggedIn()) {
+      String logoutUrl = userService.createLogoutURL(HOME_URL);
+      String nickname = getUserNickname(currentUser.getUserId());  
+
+      responseMap.put(IS_USER_LOGGED_IN, "true");
+      responseMap.put(REDIRECT_URL, logoutUrl); 
+      responseMap.put(NICKNAME, nickname); 
+
+      // Only display quotes if the user is logged in
+      String quotesJson = getQuoteListJson(request);
+      responseMap.put(QUOTE, quotesJson);
+    } else {
+      String loginUrl = userService.createLoginURL(HOME_URL);
+      responseMap.put(IS_USER_LOGGED_IN, "false");
+      responseMap.put(REDIRECT_URL, loginUrl); 
     }
 
-    String quotesJson = convertToJson(quoteList);
+    Gson gson = new Gson();
+    response.getWriter().println(gson.toJson(responseMap)); 
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("application/json");
-    response.getWriter().println(quotesJson); 
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
+      User currentUser = userService.getCurrentUser(); 
+      putQuoteIntoDatastore(request, currentUser);
+    } 
+
+    response.sendRedirect("/index.html");
```

Comment:
Maybe we can make "/index.html" a constant as well.

Reply:
Done.